signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+        if isinstance(data, Mapping):
+            return JSONDict(filename=filename, data=data, parent=parent)
+        elif isinstance(data, Sequence) and not isinstance(data, str):
+            return JSONList(filename=filename, data=data, parent=parent)
+        elif NUMPY:
+            if isinstance(data, numpy.number):
+                return data.item()
+            elif isinstance(data, numpy.ndarray):
+                return JSONList(filename=filename, data=data.tolist(), parent=parent)
+        return data
+
+    @abstractmethod
+    def to_base(self):
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    @abstractmethod
+    def _load(self):
+        pass
+
+    @abstractmethod
+    def _sync(self):
+        pass
+
+    @contextmanager
+    def _safe_sync(self):
+        backup = self._data
+        try:
+            yield
+        except BaseException:
+            self._data = backup
+            raise
+
+    def sync(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._sync()
+            else:
+                self._parent.sync()
+
+    def load(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                data = self._load()
+                with self._suspend_sync():
+                    self.reset(data)
+            else:
+                self._parent.load()
+
+
+class _SyncedDict(SyncedCollection, MutableMapping):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
+
+    VALID_KEY_TYPES = (str, int, bool, type(None))
+
+    def __init__(self, **kwargs):
+        data = kwargs.pop('data', None)
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = {}
+        else:
+            self._data = {
+                self._validate_key(key): self.from_base(data=value, parent=self)
+                for key, value in data.items()
+            }
+
+    def to_base(self):
+        converted = {}
+        for key, value in self._data.items():
+            if isinstance(value, SyncedCollection):
+                converted[key] = value.to_base()
+            else:
+                converted[key] = value
+        return converted
+
+    def reset(self, data=None):
+        if data is None:
+            data = {}
+        backup = self._data
+        if isinstance(data, Mapping):
+            try:
+                with self._suspend_sync():
+                    for key in data:
+                        if key in self._data:
+                            if data[key] == self._data[key]:
+                                continue
+                            try:
+                                self._data[key].reset(key)
+                                continue
+                            except (ValueError, AttributeError):
+                                pass
```

Comment:
Why do we need to do this reset? Once we reassign `self._data[key]` in the next line, Python should automatically clean the old object up for us.

Reply:
This gives `reset` the behavior of `dfs_update`. It doesn't change the instance for the underlying object until it is required.
It also optimizes it, as it doesn't create instances.

Reply:
Actually on further reading, I'm not sure I follow what the code is trying to accomplish. I don't think `reset` should behave like `dfs_update`, though. Correct me if I'm misunderstanding, but currently this is what the code is doing:
```
for key in data:
    if key in self._data:
        # Do the reset if data[key] is resettable, reset its value to key, then continue
    # set data[key]
```
Passing `key` as the argument to reset means that if `self._data['foo'] == {'bar': 1}` and `data['foo'] == 1`, the result of `reset` will be a `ValueError` because you're calling it with a non-Mapping, right? But even if that wasn't an error, the result would be that `self._data['foo'] == 'foo'`. Neither of these seem like the desired outcome.

I think perhaps we need to clarify the intended semantics of `reset`. I assumed this function would simply erase the contents of the current object, and if `data` is provided, load all the contents of `data` into `self._data` afterward. Is that not the intent?

Reply:
Sorry, There a typo. It should be
```
 self._data[key].reset(data[key])
```

Reply:
If `self._data['foo'] = {'bar': 1}` and `data['foo']=1`, then `reset` will return `ValueError` and it `self._data['foo'] = from_dict(data['foo'])` which is `self._data['foo'] = data['foo']` in this case

Reply:
> I don't think reset should behave like dfs_update, though.
I will make separate functions for both.

Reply:
@vishav1771 Please resolve this conversation if it has been handled completely.